### PR TITLE
Reduce broadcasted inputs in derivative code

### PR DESCRIPTION
**Pull request by apaszke**
Previously symbolic AD formulas assumed that no broadcasting happened,
and would return gradients of incorrect shapes (possibly leading to
silent errors later).

Fixes a few bugs (known and unknown):
- #11736
- ArgumentSpec didn't compute the input types correctly [(it didn't advance the offset for non-tensor args)](https://github.com/pytorch/pytorch/pull/14485/files#diff-4fd3157a056596aefb8cdf41022a208bR153)
- Symbolic AD could suffer from use after free (dangling pointers in grad map), because [`EliminateDeadCode` could have removed nodes](https://github.com/pytorch/pytorch/pull/14485/files#diff-25d33ad1ed6855684dec79d927ca6142L781) that referenced gradients of certain values.
- Undefined behavior in `aten::size`

During my tests I've also found a few new problems, and I have opened issues for them:
- FusionGroup seems to think that cat nodes broadcast their inputs (#14483)
- `prim::ConstantChunk` derivative formula doesn't handle undefined inputs (#14484)

This patch unfortunately deoptimizes some of our code (Fusion doesn't happen past chunk nodes, and outputs more tensors only because we have to get their size). I know how to fix those issues, but wanted to fix this terrible bug quickly.

cc @zou3519 @zdevito @ngimel

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -350,6 +352,17 @@ RegisterOperators reg({
             return 0;
           };
         }),
+    Operator(
+        "prim::SumToSize(Tensor(a) self, int[] size) -> Tensor(a)",
+        [](const Node* node) {
+          return [=](Stack& stack) {
+            at::Tensor self;
+            Shared<IntList> desired_sizes;
```
Comment (zou3519) [inline torch/csrc/jit/register_prim_ops.cpp]: What is `Shared<...>`?

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -775,16 +799,16 @@ Gradient differentiate(std::shared_ptr<Graph>& graph) {
...
   // Fills in f, df, f_real_outputs, df_input_captures,
   // modifies df_input_vjps (new vjps are added for temporaries)
   lambdaLiftReverse(grad_desc, rev_info);
   // It's possible the we've cloned the same constants many times, so
   // de-duplicate them
   ConstantPooling(grad_desc.df);
+  // addReverseInline has to call gradientForNode if *any* of the inputs
+  // require grad, but it will emit vjps for *all* inputs. Use DCE to remove
+  // unnecessary nodes.
+  EliminateDeadCode(grad_desc.df);
```
Comment (zou3519) [inline torch/csrc/jit/autodiff.cpp]: nit: would be good to have an extra comment about why this should happen after lambdaLiftReverse (to prevent use after free)

File: torch/csrc/jit/autodiff.cpp [hunk 3]
```diff
@@ -137,33 +138,48 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
...
-      return {grads.at(0) * inputs.at(1), grads.at(0) * inputs.at(0)};
+      return {sumToSizeOf(grads.at(0) * inputs.at(1), attr::self),
+              sumToSizeOf(grads.at(0) * inputs.at(0), attr::other)};
 
     } else if (node->matches("aten::mul(Tensor self, Scalar other) -> Tensor")) {
       return {grads.at(0) * inputs.at(1), nullptr};
 
     } else if (node->matches("aten::div(Tensor self, Tensor other) -> Tensor")) {
-      return {grads.at(0) / inputs.at(1), -grads.at(0) * inputs.at(0) / (inputs.at(1) * inputs.at(1))};
+      return {sumToSizeOf(grads.at(0) / inputs.at(1), attr::self),
```
Comment (zou3519) [inline torch/csrc/jit/autodiff.cpp]: (putting this comment here because github won't let me comment where I want to)
max and min also broadcast, so those need `sumToSize` as well

Comment (ngimel) [reply torch/csrc/jit/autodiff.cpp]: I thought max and min insert explicit expands https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L1382-L1383 and expands are handled in backward? I'm also wondering if it would be better to insert expands in forward where needed, rather than unconditionally put sumToSize everywhere?

Comment (zou3519) [reply torch/csrc/jit/autodiff.cpp]: I'm not sure -- I don't see explicit expands in
```
import torch

def fnsum(a, b):
    a = a.unsqueeze(2)
    b = b.unsqueeze(1)
    sum = torch.max(a, b)
    out = torch.tanh(sum)
    return out

if __name__ == '__main__':
    a1 = torch.randn(2, 3, 5, requires_grad=True)
    a2 = torch.randn(2, 4, 5, requires_grad=True)
    fn = torch.jit.script(fnsum)
    out = fn(a1, a2)
    print(fn.graph_for(a1, a2))
    out.backward(torch.ones_like(out))
```

File: torch/csrc/jit/register_prim_ops.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/jit/register_prim_ops.cpp]: It's some kind of shared pointer that's implemented in this file. All list functions use code like this (search for e.g. `listNe`)

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Oh yes I had it but must have removed it accidentally

File: torch/csrc/jit/autodiff.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: @ngimel Script never inserts explicit expands.

Comment (ngimel) [reply torch/csrc/jit/autodiff.cpp]: Ideally, there could be a middle ground between hard-coded sizes and promises like "there won't be a broadcast in this op" or "expect dim x of tensor y to be broadcasted". In many usecases, tensor sizes can be different, but dimensionality and contiguity stay the same.

Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Dimensionality (in terms of the number of dimensions) and contiguity yes, but those are insufficient guarantees to prove that no broadcasts will happen.

Comment (ngimel) [reply torch/csrc/jit/autodiff.cpp]: Symbolic expressions might be that happy middle ground!

**Comment by apaszke** [comment]
Turns out the ASAN failure was triggered by UB in the existing implementation of `aten::size`. Should be fixed now.

File: test/cpp/jit/tests.h [hunk 4]
```diff
@@ -867,11 +866,11 @@ void testDifferentiateWithRequiresGrad(std::ostream& out = std::cout) {
   PropagateRequiresGrad(graph);
 
   auto grad_spec = differentiate(graph);
-  std::vector<size_t> expected_input_vjps = {1, 2}; // for e and %4 = (d + a)
+  std::vector<size_t> expected_input_vjps = {1, 4}; // for e and %4 = (d + a)
```
Comment (zou3519) [inline test/cpp/jit/tests.h]: The extra 2 are from saving the sizes, right? It might be good to write that in a comment here

File: test/test_jit.py [hunk 5]
```diff
@@ -351,7 +351,7 @@ def allSum(vs):
                 vs = vs[:-drop]
             # we don't want all the grad for all the outputs to be the same
             # so we multiply each by a constant
-            return sum([(i + 1) * v.sum() for i, v in enumerate(vs) if v is not None])
+            return sum([math.log(i + 2) * v.sum() for i, v in enumerate(vs) if v is not None])
```
Comment (zou3519) [inline test/test_jit.py]: I'm not sure what is going on here: why did this need to change?

File: test/test_jit.py [hunk 6]
```diff
@@ -426,10 +426,11 @@ def allSum(vs):
 
         return ge
 
-    def assertAllFused(self, graph):
+    def assertAllFused(self, graph, except_for=()):
```
Comment (zou3519) [inline test/test_jit.py]: nit: maybe `except_for=frozenset()`? Since sets get passed in for except_for

File: test/cpp/jit/tests.h (hunk 4 above)
Comment (apaszke) [reply test/cpp/jit/tests.h]: It's only an increased offset, the comment in here is up to date. I'm not sure if it's worth adding a comment there, since it's going to be the standard practice to carry over a ton of sizes to backward from now on.

File: test/test_jit.py (hunk 5 above)
Comment (apaszke) [reply test/test_jit.py]: Because `(i + 1)` became really large when we had tensors with thousands of outputs, and that caused the second order gradients to be in the order of tens of thousands at least. That's not a good regime for floating-point calculations, so I replaced it with a much slower growing `log`.

File: test/test_jit.py (hunk 6 above)
Comment (apaszke) [reply test/test_jit.py]: Sure, I could change that, but technically it doesn't have to be a set (only an iterable), since we'll call `set(except_for)` anyway.

File: torch/csrc/autograd/engine.cpp [hunk 7]
```diff
@@ -340,7 +340,7 @@ static void validate_outputs(const edge_list& edges, variable_list& grads, const
         ss << metadata.shape();
         AT_ERROR(format_error(ss.str()));
       }
-      grads[i] = at::sum_to(grads[i], metadata.shape());
+      grads[i] = at::sum_to(std::move(grads[i]), metadata.shape());
```
Comment (zou3519) [inline torch/csrc/autograd/engine.cpp]: Is `grads[i]` a value or a reference? `grads` (the input to this function) gets reused after the invocation to this function (line 381), so does this invalidate the entry?

Comment (apaszke) [reply torch/csrc/autograd/engine.cpp]: `grads[i]` is a reference, because otherwise it would already be an rvalue, and the `std::move` would be unnecessary. It invalidates the entry for the time of calling into `sum_to`, and replaces it immediately after, so I think it preserves the current behavior.

File: test/test_jit.py (hunk 6 above)
Comment (zou3519) [reply test/test_jit.py]: Oh I see, that's fine then, thanks for the clarification

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Actually I've realized that we absolutely need this pass in the place where it was called previously, or it won't clean up all the code correctly (ok, it will, but we will end up with unused captures between stages). I think it's best to merge this as is, and I'll send a new PR that fixes this issue, and adds `CommonSubexpressionElimination`, because at the moment we can emit a lot of `aten::size` calls on a given tensor, and capture all of them as if they were independent values.

File: torch/csrc/jit/autodiff.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Fixed.

**Comment by apaszke** [comment]
Alright, I'd like to get this in ASAP so that I can rebase my fuser PR on top of it. I'll also send another PR with a cleanup of DCE + CSE for the primal/reverse blocks, because the code we emit right now ends up being very messy.

**Comment by ailzhang** [comment]
@apaszke  sure I will land it once the CI passes. (or do you want to skip it instead? :P

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (zou3519) [reply torch/csrc/jit/autodiff.cpp]: Sounds good, go forth

**Comment by apaszke** [comment]
@ailzhang please wait for an approving review before landing

File: test/expect/TestScript.test_milstm_fusion_cuda-backward.expect [hunk 8]
```diff
@@ -20,87 +20,121 @@ graph(%0 : Float(*, *)
...
+  %3 : Float(*, *) = aten::mul(%2, %2)
+  %4 : Float(*, *) = aten::neg(%3)
+  %5 : int = prim::Constant[value=1]()
+  %6 : Float(*, *) = aten::add(%4, %5, %5)
+  %7 : Dynamic = aten::mul(%1, %6)
+  %8 : int = prim::Constant[value=1]()
+  %9 : Dynamic = aten::add(%0, %7, %8)
+  return (%9);
+}
+with prim::FusionGroup_1 = graph(%0 : Dynamic
```
Comment (zou3519) [inline test/expect/TestScript.test_milstm_fusion_cuda-backward.expect]: There were three fusion groups before but now it turned into two -- that is interesting

File: test/expect/TestJit.test_cpp_cuda.expect [hunk 9]
```diff
@@ -135,35 +153,53 @@ graph(%0 : Float(*)
   %3 : int = prim::Constant[value=1]()
   %4 : Float(*) = aten::add(%2, %1, %3)
   %5 : int = prim::Constant[value=1]()
+  %26 : int[] = aten::size(%4)
```
Comment (zou3519) [inline test/expect/TestJit.test_cpp_cuda.expect]: nit: It might be good to add @ngimel's example from #11736 into test/test_jit.py to have a non-expect-test that checks the derivatives of a broadcasting op are okay.

**Review by zou3519 — APPROVED** [review]
Code looks correct. The deoptimization is unfortunate but at least we have ideas on how to fix it soon

File: test/expect/TestScript.test_milstm_fusion_cuda-backward.expect (hunk 8 above)
Comment (apaszke) [reply test/expect/TestScript.test_milstm_fusion_cuda-backward.expect]: Huh, that's surprising! One fusion group in LSTM got split into two instead.

File: test/expect/TestJit.test_cpp_cuda.expect (hunk 9 above)
Comment (apaszke) [reply test/expect/TestJit.test_cpp_cuda.expect]: Good point, I forgot to do that, even though I tested this solution on that snippet. Will do that in my followup PR that cleans up the DCE situation.

**Comment by ailzhang** [comment]
@apaszke @zou3519 If nothing will be added to this PR, I will land it soon?

**Comment by apaszke** [comment]
@ailzhang unfortunately the PR got into some conflicts with master, but they should be resolved now, so please reimport and land whenever you can!

**Comment by ailzhang** [comment]
FYI landing failed twice due to `Your commit was blocked by a pre-commit hook.` which should be fixed by a rebase.

**Comment by ailzhang** [comment]
Weird... I kept getting land failed for this PR. @soumith  @zou3519 Could you help take a look when you have time? Thanks!

**Comment by zou3519** [comment]
I'll take a look this morning

**Comment by suo** [comment]
@apaszke can you push this change to your PR branch?
```
diff --git a/torch/csrc/jit/autodiff.cpp b/torch/csrc/jit/autodiff.cpp
index eb5bb2171..1e16334f7 100644
--- a/torch/csrc/jit/autodiff.cpp
+++ b/torch/csrc/jit/autodiff.cpp
@@ -809,7 +809,7 @@ Gradient differentiate(std::shared_ptr<Graph>& graph) {
   // addReverseInline has to call gradientForNode if *any* of the inputs
   // require grad, but it will emit vjps for *all* inputs. Use DCE to remove
   // unnecessary nodes.
-  EliminateDeadCode(grad_desc.df);
+  EliminateDeadCode(grad_desc.df->block());
   return grad_desc;
 }
```
